Network messages

ottojo opened the pull request:
fixes #24 

## Spielinitialisierung
- [x] HELLO,
- [x] HELLO_REPLY,
- [x] RECONNECT,
- [x] GAME_STARTED,
##  Wahlphase
- [x] REQUEST_ITEM_CHOICE,
- [x] ITEM_CHOICE,
- [x] REQUEST_EQUIPMENT_CHOICE,
- [x] EQUIPMENT_CHOICE,
##  Spielphase
- [x] GAME_STATUS,
- [x] REQUEST_GAME_OPERATION,
- [x] GAME_OPERATION,
##  Spielende
- [x] STATISTICS,
##  Kontrollnachrichten
- [x] GAME_LEAVE,
- [x] GAME_LEFT,
- [x] REQUEST_GAME_PAUSE,
- [x] GAME_PAUSE,
- [x] REQUEST_META_INFORMATION,
- [x] META_INFORMATION,
- [x] STRIKE,
- [x] ERROR,
##  Optionale Komponenten
- [x] REQUEST_REPLAY,
- [x] REPLAY

ottojo committed:
add MessageContainer

ottojo committed:
add ConfigDeliveryMessage

ottojo committed:
add GameLeaveMessage

ottojo committed:
adds GameLeftMessage

ottojo committed:
add ConnectionErrorMessage

ottojo committed:
add EquipmentChoiceMessage

ottojo committed:
add new classes to CMakeLists.txt

ottojo committed:
fixed warnings (unused argument, extra ;, initialization order)

ottojo committed:
add empty Scenario class

ottojo committed:
add MessageTypeEnum

ottojo committed:
adds OptionalSerialization.hpp

ottojo committed:
MatchConfig to_json conversion function declaration

ottojo committed:
create CharacterInformation json functions

ottojo committed:
CharacterDescription default constructor, make members non-const to not implicitly delete move/copy constructor

ottojo committed:
add GameOperationMessage

ottojo committed:
remove redundant default construction of members in CharacterInformation

ottojo committed:
CharacterDescription member initialization and default constructor

ottojo committed:
prefer class initialization to initializer list

ottojo committed:
add GamePauseMessage

ottojo committed:
move `network::messages::MessageContainer` to `network::MessageContainer`

ottojo commented:
Ist zwar noch nicht fertig, wird aber großzügig viel, review kann also gerne schon angefangen werden...

Github kann das ja ganz gut dass man nur Sachen reviewt die man noch nicht angeschaut hat.

Marco5de committed:
WIP: Hello to_json funktioniert noch nicht

ottojo committed:
Merge branch 'develop' into network_messages

ottojo committed:
Merge branch 'develop' into network_messages

ottojo committed:
fix from_json method signature in HelloMessage

ottojo committed:
fix UUID null-value in HelloMessage test

ottojo committed:
fix second UUID null-value in HelloMessage test

ottojo committed:
create directory for message tests

ottojo committed:
update EquipmentChoice to conform to standard

ottojo committed:
remove ConfigDeliveryMessage and ConnectionErrorMessage (std V2)

ottojo committed:
HelloMessage: add default Role, add to CMakeLists.txt

ottojo committed:
rename message::HelloMessage to message::Hello

ottojo committed:
update MessageTypeEnum according to standard

ottojo committed:
remove renamed sources from CMakeLists.txt

ottojo committed:
update Error messages

ottojo committed:
Add invalid type to RoleEnum

ottojo committed:
add proper constructors to GamePauseMessage

ottojo committed:
add new error types

ottojo committed:
add HelloReplyMessage

Marco5de committed:
Game started message

Marco5de committed:
reconnect message

Marco5de committed:
fixed non public inheritance

Marco5de committed:
fixed CMAKE

Marco5de committed:
request item choice message

Marco5de committed:
item choice message

Marco5de committed:
Request eq choice

Marco5de committed:
request game operation message

Marco5de committed:
game status
 message

Marco5de committed:
fixed typo

Marco5de committed:
Request game pause message

Marco5de committed:
Request Meta Information message

Marco5de committed:
strike message

Marco5de committed:
request replay message

Marco5de committed:
replay

Marco5de committed:
added statistics and corresponding message

Marco5de committed:
fixed include guards

csacro committed:
Merge branch 'develop' into network_messages

Marco5de commented on src/network/messages/Hello.cpp:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    void to_json(nlohmann::json &j, const spy::network::messages::Hello &h) {
+        MessageContainer::common_to_json(j, h);
```
> Wichtig: Diese Nachricht hält das clientId-Feld im MessageContainer frei – genauer gesagt wird
es vom Server ignoriert!

sollte also in Ordnung sein

ottojo commented on src/network/MessageContainer.cpp:
```diff
@@ -0,0 +1,21 @@
+        MessageContainer::common_from_json(j, m);
+    }
+
+    // TODO initialize creationdate
```
-> Low Prio -> issue

Marco5de commented on src/network/messages/ItemChoice.hpp:
```diff
@@ -0,0 +1,36 @@
+
+
+        private:
+            util::UUID chosenCharacter;
```
Eines der beiden Attribute wird explizit `null` gesetzt.
Sinnvolle repräsentation dafür?

ottojo replied:
Kann man bestimmt fancy mit `std::variant` machen. Ich prüfe das mal

ottojo replied:
done.

ottojo commented on src/network/MessageContainer.hpp:
```diff
@@ -0,0 +1,69 @@
+             * @tparam T Message type deriving from MessageContainer
+             */
+            template<typename T>
+            static void common_to_json(nlohmann::json &j, const T &message) {
```
das müsste doch auch mit `MessageContainer` statt `typename T` gehen, das ist vielleicht sinnvoller.

ottojo commented on src/network/MessageContainer.hpp:
```diff
@@ -0,0 +1,69 @@
+             * @tparam T Message type deriving from MessageContainer
+             */
+            template<typename T>
+            static void common_from_json(const nlohmann::json &j, T &message) {
```
siehe oben

Marco5de commented on src/network/messages/StatisticsMessage.hpp:
```diff
@@ -0,0 +1,44 @@
+        private:
+            util::UUID winner;
+            spy::statistics::VictoryEnum reason;
+            spy::statistics::Statistics statistics;
```
Feld ist optional und sollte auch als solches implementiert werden.

ottojo commented on src/util/OptionalSerialization.hpp:
```diff
@@ -0,0 +1,38 @@
+
+#include <nlohmann/json.hpp>
+
+// TODO less criminal C++ (thou shalt not hack the std namespace)
```
Das müsste auf develop schon gefixt sein -> merge

ottojo committed:
Merge branch 'develop' into network_messages

ottojo committed:
remove duplicate code blocks resulting from merge

ottojo committed:
Use parent class instead of template for MessageContainer::common_to_json and MessageContainer::common_from_json

ottojo committed:
ErrorTypeEnum: Keep "invalid" entry for parsing failures

ottojo committed:
Fix various codesmells

ottojo committed:
use std::variant for item choice that can either be Character or Gadget

ottojo committed:
make StatisticsMessage::statistics optional

csacro commented on src/datatypes/character/CharacterInformation.hpp:
```diff
@@ -30,6 +30,9 @@ namespace spy::character {
 
             friend void from_json(const nlohmann::json &j, CharacterInformation &ci);
 
+            friend void to_json(nlohmann::json &j, const CharacterInformation &c);
+
+            friend void from_json(const nlohmann::json &j, CharacterInformation &c);
```
duplicat code
diese Methoden sollten gelöscht werden (die zwei analogen Methoden darüber wurden implementiert)

authaldo replied:
Eventuell merge Überbleibsel? to_json und from_json gibts doch schon?

ottojo committed:
Merge branch 'develop' into network_messages

ottojo committed:
remove duplicate json method declarations from CharacterInformation.hpp

authaldo commented on src/network/messages/Replay.hpp:
```diff
@@ -0,0 +1,83 @@
+            util::UUID playerTwoId;
+            std::string playerOneName;
+            std::string playerTwoName;
+            int rounds;
```
Warum nicht unsigned?

```suggestion
            unsigned int rounds;
```

ottojo replied:
Gute idee, mach ich

authaldo commented on src/network/messages/Strike.hpp:
```diff
@@ -0,0 +1,37 @@
+            friend void from_json(const nlohmann::json &j, Strike &s);
+
+        private:
+            int strikeNr;
+            int strikeMax;
```
Hier könnte man auch über unsigned nachdenken, bisher steht im Standard noch nichts davon, dass es für strike max eine infinte Option (bspw. über negative Werte) geben muss.

ottojo replied:
Würde ich so lassen aufgrund deiner Argumente.

authaldo reviewed (changes requested):
Ich bin generell Fan von Doxygen Headern und Kommentaren, aber die kann man auch noch im Verlauf ergänzen. Außerdem finde ich es komisch, dass die Kommentare teilweise auf deutsch sind (z.b.MessageTypeEnum.hpp), vielleicht könnte man da einheitlich bei Englisch bleiben.

ottojo committed:
make rounds unsigned in Replay Message

ottojo commented:
Vorschlag zur Coverage: Ich bastel gerade was, das mit wenig Aufwand pro Typ eine Beispielnachricht encoded und wieder decoded. Das sollte wenigstens Konsistenz sichern. Falls es in diesem Hinblick später Probleme gibt, kann man ja immer noch Testdaten generieren + nutzen, oder schemas bauen und so.

Dominik hat vorgeschlagen das in neuem PR zu machen, finde ich OK.

csacro commented on src/network/messages/HelloReply.hpp:
```diff
@@ -0,0 +1,37 @@
+
+
+        private:
+            util::UUID sessionId;
+            scenario::Scenario level;
+            MatchConfig settings;
+            std::vector<character::CharacterInformation> characterSettings;
```
how to access fields -> getter

csacro commented on src/network/MessageContainer.hpp:
```diff
@@ -0,0 +1,48 @@
+            static void common_from_json(const nlohmann::json &j, MessageContainer &message);
+
+        private:
+            spy::util::UUID playerId;
+            messages::MessageTypeEnum type = messages::MessageTypeEnum::INVALID;
+            // TODO: std::chrono::system_clock::time_point creationDate;
+            std::string creationDate;
+            std::optional<std::string> debugMessage;
```
how to access fields -> getter
für die `debugMessage` wäre noch ein setter gut

csacro commented on src/network/ErrorTypeEnum.hpp:
```diff
@@ -0,0 +1,30 @@
+
+namespace spy::network {
+    enum class ErrorTypeEnum {
+            INVALID,
```
Im Standard gibt es noch `GENERAL` im `ErrorTypeEnum`. Man könnte sich auch überlegen das als default statt `INVALID` zu nehmen

csacro commented on src/network/messages/RequestEquipmentChoice.hpp:
```diff
@@ -0,0 +1,35 @@
+
+            RequestEquipmentChoice();
+
+            explicit RequestEquipmentChoice(const util::UUID &playerId);
```
Sollte man nicht im Konstruktor auch `chosenCharacterIds` und `chosenGadgets` setzen können ?

csacro commented on src/network/messages/RequestItemChoice.hpp:
```diff
@@ -0,0 +1,35 @@
+
+            RequestItemChoice();
+
+            explicit RequestItemChoice(const util::UUID &playerId);
```
Sollte man im Konstruktor nicht auch `offeredCharacterIds` und `offeredGadgets` setzen können ?

csacro commented on src/network/messages/EquipmentChoice.hpp:
```diff
@@ -0,0 +1,31 @@
+
+            EquipmentChoice();
+
+            explicit EquipmentChoice(util::UUID playerId);
```
Sollte man im Konstruktor nicht `equipment` setzen können ?

csacro commented on src/network/messages/GameOperation.hpp:
```diff
@@ -0,0 +1,28 @@
+            friend void from_json(const nlohmann::json &j, GameOperation &g);
+
+        private:
+            gameplay::Operation operation;
```
how to acces `operation` -> getter

csacro commented on src/network/messages/GamePause.hpp:
```diff
@@ -0,0 +1,28 @@
+            friend void from_json(const nlohmann::json &j, GamePause &g);
+
+        private:
+            bool gamePause = false;
```
fehlendes Feld: `boolean serverEnforced`

csacro commented on src/datatypes/statistics/VictoryEnum.hpp:
```diff
@@ -0,0 +1,37 @@
+#include <nlohmann/json.hpp>
+
+namespace spy::statistics {
+    enum class VictoryEnum {
+            VICTORY_BY_IP,
+            VICTORY_BY_COLLAR,
+            VICTORY_BY_DRINKING,
+            VICTORY_BY_SPILLING,
+            VICTORY_BY_HP,
+            VICTORY_BY_RANDOMNESS,
+            VICTORY_BY_LEAVE,
+            VICTORY_BY_KICK
```
Sollen wir hier auch noch einen extra default `INVALID` oder so machen ?

csacro reviewed (changes requested):
`MetaInformation.hpp` und `MetaInformation.cpp` fehlen (`MetaInformationMessage` page 42 im aktuellen Standard).

csacro commented:
> 
> 
> Vorschlag zur Coverage: Ich bastel gerade was, das mit wenig Aufwand pro Typ eine Beispielnachricht encoded und wieder decoded. Das sollte wenigstens Konsistenz sichern. Falls es in diesem Hinblick später Probleme gibt, kann man ja immer noch Testdaten generieren + nutzen, oder schemas bauen und so.
> 
> Dominik hat vorgeschlagen das in neuem PR zu machen, finde ich OK.

Geht von mir aus klar. Wenn sonst alles passt werde ich das mergen. Issue ist bereits erstellt und dem Sprint hinzugefügt. Habe mal dich (@ottojo ) assigned, weil du meinstest du bist schon dran.

ottojo committed:
Review from @csacro

add field serverEnforced to GamePauseMessage

add default INVALID type to VictoryEnum

add getter for GameOperation::operation

add equipment field of EquipmentChoice to constructor

add members to constructor for RequestItemChoiceMessage

add members to constructor for RequestEquipmentChoiceMessage

add GENERAL error type

add getter and setter for MessageContainer

add getters for HelloReplyMessage

ottojo committed:
add MetaInformationMessage
